Enable Yii migrations after import of production database backup for #792

rija opened the pull request:
The Yii schema migrations fail to run after we import a database backup because the subjects of each migration already exist.
The solution is three fold:

* Ensure existing migrations that create tables, sequences, views only proceed if the object doesn't exist yet
* Drop all constraints before running migrations using a Yii command
* Drop all indexes before running migrations using a Yii command

Implementation:

* [x] Migration to create table only if not exists
* [x] Migration to create sequence only if not exists 
* [x] Migration to create view only if not exists
* [x] The GitLab pipeline deployment job has been updated to run the constraints dropping command before running the schema migrations.
* [x] The GitLab pipeline deployment job has been updated to run the index dropping command before running the schema migrations.
* [x] docs are updated

rija committed:
Add IF NOT EXISTS clause to CREATE TABLE clauses in schema migration

rija committed:
Add IF NOT EXISTS clause to CREATE SEQUENCE clauses in schema migration

rija committed:
Add OR REPLACE clause to CREATE VIEW clauses in schema migration

to avoid errors if the view already exists

rija committed:
Add Yii command to drop all constraints and indexes in the database

rija committed:
Update docs to mention the new Yii migrations pre-requisites

Update the README's database migration sections
Update the sql/README's section on how to use the newly converted database dump

rija committed:
Re-enable file type and format now that migrations works

kencho51 reviewed (approved):
I could deploy this PR to [staging server](https://ec2-live.gigadb.link/) and [live server](https://ec2-live.gigadb.link/), with the manual migration of the latest production database downloaded using `files-url-updater` tool, all datasets are available in both servers.

pli888 commented:
I managed to upgrade the GigaDB PostgreSQL database with the new code described in this PR to my local dev machine and AWS staging server. There was one strange bug I had to fix to get everything working - this step did not work for me in the `convert_production_db_to_latest_ver.sh` script:
```
$ ./ops/scripts/convert_production_db_to_latest_ver.sh
Download and load the production database in to postgreSQL server using file-url-updater
version: 9.3.25
latest: 20210921

Command is running for date 20210921

Downloading production backup for 20210921
/app/sql/gigadbv3_20210921.backup:                      30.09 MB    6.65 MB/s  
/app/sql/gigadbv3_20210921.backup:                      30.09 MB    6.65 MB/s  

Warning! This command will drop the configured database (hosted on pg9_3) and restore it from the 20210921 backup, are you sure you want to proceed?
 (yes|no) [no]:yes
Executing command...

Restoring the backup for 20210921
Export production data as text (only the strictly necessary data is exported)
pg_dump dumps a database as a text file or to other formats.

Usage:
  pg_dump [OPTION]... [DBNAME]

General options:
  -f, --file=FILENAME          output file or directory name
  -F, --format=c|d|t|p         output file format (custom, directory, tar,
                               plain text (default))
  -j, --jobs=NUM               use this many parallel jobs to dump
  -v, --verbose                verbose mode
  -V, --version                output version information, then exit
  -Z, --compress=0-9           compression level for compressed formats
  --lock-wait-timeout=TIMEOUT  fail after waiting TIMEOUT for a table lock
  -?, --help                   show this help, then exit
<snip></snip>
If no database name is supplied, then the PGDATABASE environment
variable value is used.

Report bugs to <pgsql-bugs@postgresql.org>.
No upgraded database found, conversion fail!
```

The way I got this shell script to run was to make this change:
```
echo "Export production data as text (only the strictly necessary data is exported)"
$DOCKER_COMPOSE run --rm updater pg_dump --host=pg9_3 --username=gigadb --clean --create --schema=public --no-privileges --no-tablespaces --dbname=gigadb --file=sql/gigadbv3_"$latest"_v"$version".backup
```

I wasn't able to run the code to deploy a production database backup onto a live server because AWS was complaining that we had reached the maximum of 5 elastic IP addresses allowed. I assume live deployment will work in the same way as staging deployment so I approve this PR.

pli888 reviewed (approved):
I managed to upgrade the GigaDB PostgreSQL database with the new code described in this PR to my local dev machine and AWS staging server. There was one strange bug I had to fix to get everything working - this step did not work for me in the `convert_production_db_to_latest_ver.sh` script:
```
$ ./ops/scripts/convert_production_db_to_latest_ver.sh
Download and load the production database in to postgreSQL server using file-url-updater
version: 9.3.25
latest: 20210921

Command is running for date 20210921

Downloading production backup for 20210921
/app/sql/gigadbv3_20210921.backup:                      30.09 MB    6.65 MB/s  
/app/sql/gigadbv3_20210921.backup:                      30.09 MB    6.65 MB/s  

Warning! This command will drop the configured database (hosted on pg9_3) and restore it from the 20210921 backup, are you sure you want to proceed?
 (yes|no) [no]:yes
Executing command...

Restoring the backup for 20210921
Export production data as text (only the strictly necessary data is exported)
pg_dump dumps a database as a text file or to other formats.

Usage:
  pg_dump [OPTION]... [DBNAME]

General options:
  -f, --file=FILENAME          output file or directory name
  -F, --format=c|d|t|p         output file format (custom, directory, tar,
                               plain text (default))
  -j, --jobs=NUM               use this many parallel jobs to dump
  -v, --verbose                verbose mode
  -V, --version                output version information, then exit
  -Z, --compress=0-9           compression level for compressed formats
  --lock-wait-timeout=TIMEOUT  fail after waiting TIMEOUT for a table lock
  -?, --help                   show this help, then exit
<snip></snip>
If no database name is supplied, then the PGDATABASE environment
variable value is used.

Report bugs to <pgsql-bugs@postgresql.org>.
No upgraded database found, conversion fail!
```

The way I got this shell script to run was to make this change:
```
echo "Export production data as text (only the strictly necessary data is exported)"
$DOCKER_COMPOSE run --rm updater pg_dump --host=pg9_3 --username=gigadb --clean --create --schema=public --no-privileges --no-tablespaces --dbname=gigadb --file=sql/gigadbv3_"$latest"_v"$version".backup
```

I wasn't able to run the code to deploy a production database backup onto a live server because AWS was complaining that we had reached the maximum of 5 elastic IP addresses allowed. I assume live deployment will work in the same way as staging deployment so I approve this PR.

rija committed:
Pass DB name using --dbname parameter for some version of pg_dump

rija committed:
Add GitLab button to reset migration to zero

rija committed:
Update database conversion script to pass custom date

sometimes the production backup is corrupted so we need the ability to specify
a specific date

rija commented:
@pli888 
Thanks, I've changed the script to use the ``--dbname`` parameter and it works fine.
Additionally, I had to update the database conversion script to accept custom date, not just the latest, because the CNGB backup for 20210922 seems to be corrupted (the sample table is empty). In such circumstances we would want to use a backup from a different date for the conversion.
